mqtt/mqtt-bridge/src/persist/loader.rs

```diff
@@ -78,9 +77,34 @@ where
 
             // refresh next batch
             // if error, either someone forged the database or we have a database schema change
-            let next_batch = self.next_batch()?;
+            let mut new_batch = self.next_batch()?;
+
+            // drop those loaded keys which do not exist in the new batch
             let mut inner = self.0.lock();
-            inner.batch = next_batch;
+            if let Some((new_key, _)) = new_batch.front() {
+                while inner.loaded.front().map_or(false, |key| new_key != key) {
+                    inner.loaded.pop_front();
+                }
+            }
+
+            // drop those items from a new batch that were loaded but not removed from storage
+            if !new_batch.is_empty() {
+                for key in &inner.loaded {
+                    match new_batch.front() {
+                        Some((new_key, _)) if new_key == key => {
+                            new_batch.pop_front();
+                        }
+                        _ => panic!(
+                            "Invalid MessageLoader state: new_batch is corrupted {:?} for a key {}",
+                            new_batch, key
```

Comment:
Do we want to panic or should we bubble up an error?

Reply:
I decided to fail fast here because I rely underlying storage is not corrupted.

Reply:
will this crash the process? If so, then this is probably not a good idea. I don't think our storage has been hardened yet.

Reply:
Yes, it will crash the process and we will go and investigate the issue. Do you suggest we log an error and continue?

Reply:
I am okay with crashing, as the RingBuffer **should** be able to come back up where it left off and restart the batch, hopefully success on the second time. But maybe the message about corruption could be changed, as that makes it sound like data corruption to me, and I feel this problem is a bit different.

Reply:
Yes, crashing is probably not a good idea at this point - it will be reasonable if this is never expected to happen, but as I said, the store is not hardened enough for that. Also, we don't really collect crash dumps, etc. which will help us analyze the reason why it went into this state. 
It also might increase the potential of it causing further corruption of the store state (and possibly cause us to lose logs.. not tested this though)